File: _sandbox/tests/test_ui_state.py
import concurrent.futures
import threading
import time

from ui_state import UIState


class DummySelect:
    def __init__(self):
        self.options = []
        self.update_called = 0

    def update(self):
        self.update_called += 1


def test_concurrent_clear_and_append():
    state = UIState({'chat_container': [], 'chat_history': []})

    def worker(i):
        # alternate between append and clear to provoke races
        if i % 3 == 0:
            state.clear_chat()
        else:
            state.append_chat_message(i)

    # Run many concurrent workers
    with concurrent.futures.ThreadPoolExecutor(max_workers=16) as ex:
        futures = [ex.submit(worker, i) for i in range(200)]
        for f in concurrent.futures.as_completed(futures, timeout=10):
            f.result()

    # After concurrent ops, ensure structures are consistent and same length
    cont = state.get('chat_container')
    hist = state.get('chat_history')

    assert isinstance(cont, list)
    assert isinstance(hist, list)
    assert len(cont) == len(hist)
    # All items must be integers (we only appended ints)
    assert all(isinstance(x, int) for x in cont)
    assert all(isinstance(x, int) for x in hist)


def test_update_model_options_calls_update():
    sel = DummySelect()
    state = UIState({'model_select': sel})

    state.update_model_options(['a', 'b', 'c'])

    assert sel.options == ['a', 'b', 'c']
    assert sel.update_called >= 1

File: rust_output/src/test_ui_state.rs
use anyhow::{Result, Context};
use crate::ui_state::{UIState};
use std::collections::HashMap;

/// DummySelect class.
#[derive(Debug, Clone)]
pub struct DummySelect {
    pub options: Vec<serde_json::Value>,
    pub update_called: i64,
}

impl DummySelect {
    pub fn new() -> Self {
        Self {
            options: vec![],
            update_called: 0,
        }
    }
    pub fn update(&mut self) -> () {
        self.update_called += 1;
    }
}

/// Test concurrent clear and append.
pub fn test_concurrent_clear_and_append() -> () {
    // Test concurrent clear and append.
    let mut state = UIState(HashMap::from([("chat_container".to_string(), vec![]), ("chat_history".to_string(), vec![])]));
    let worker = |i| {
        // Worker.
        if (i % 3) == 0 {
            state::clear_chat();
        } else {
            state::append_chat_message(i);
        }
    };
    let mut ex = concurrent.futures.ThreadPoolExecutor(/* max_workers= */ 16);
    {
        let mut futures = 0..200.iter().map(|i| ex.submit(worker, i)).collect::<Vec<_>>();
        for f in concurrent.futures.as_completed(futures, /* timeout= */ 10).iter() {
            f.result();
        }
    }
    let mut cont = state::get(&"chat_container".to_string()).cloned();
    let mut hist = state::get(&"chat_history".to_string()).cloned();
    assert!(/* /* isinstance(cont, list) */ */ true);
    assert!(/* /* isinstance(hist, list) */ */ true);
    assert!(cont.len() == hist.len());
    assert!(cont.iter().map(|x| /* /* isinstance(x, int) */ */ true).collect::<Vec<_>>().iter().all(|v| *v));
    assert!(hist.iter().map(|x| /* /* isinstance(x, int) */ */ true).collect::<Vec<_>>().iter().all(|v| *v));
}

/// Test update model options calls update.
pub fn test_update_model_options_calls_update() -> () {
    // Test update model options calls update.
    let mut sel = DummySelect();
    let mut state = UIState(HashMap::from([("model_select".to_string(), sel)]));
    state::update_model_options(vec!["a".to_string(), "b".to_string(), "c".to_string()]);
    assert!(sel.options == vec!["a".to_string(), "b".to_string(), "c".to_string()]);
    assert!(sel.update_called >= 1);
}
